ksqldb-engine/src/main/java/io/confluent/ksql/execution/pull/StreamedRowTranslator.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Copyright 2022 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"; you may not use
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.execution.pull;
+
+import com.google.common.base.Preconditions;
+import io.confluent.ksql.rest.entity.KsqlHostInfoEntity;
+import io.confluent.ksql.rest.entity.StreamedRow;
+import io.confluent.ksql.rest.entity.StreamedRow.Header;
+import io.confluent.ksql.schema.ksql.LogicalSchema;
+import io.confluent.ksql.util.ConsistencyOffsetVector;
+import io.confluent.ksql.util.KsqlException;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+public class StreamedRowTranslator implements Iterator<PullQueryRow> {
+
+  private final Deque<PullQueryRow> rowQueue = new ArrayDeque<>();
+  private final LogicalSchema expectedSchema;
+  private final Optional<ConsistencyOffsetVector> offsetVector;
+
+  private Header header = null;
+
+  public StreamedRowTranslator(
+      final LogicalSchema expectedSchema,
+      final Optional<ConsistencyOffsetVector> offsetVector
+  ) {
+    this.expectedSchema = expectedSchema;
+    this.offsetVector = offsetVector;
+  }
+
+  @Override
+  public PullQueryRow next() {
+    return rowQueue.pollFirst();
+  }
+
+  @Override
+  public boolean hasNext() {
+    return !rowQueue.isEmpty();
+  }
+
+  public void add(final List<StreamedRow> rows) {
+    if (rows == null || rows.isEmpty()) {
+      return;
+    }
+
+    for (int i = 0; i < rows.size(); i++) {
+      final StreamedRow row = rows.get(i);
+      if (row.getHeader().isPresent()) {
+        handleHeader(row);
+      } else if (row.getErrorMessage().isPresent()) {
+        // If we receive an error that's not a network error, we let that bubble up.
+        throw new KsqlException(row.getErrorMessage().get().getMessage());
+      } else if (!row.getRow().isPresent()) {
+        handleNonDataRows(row, i, offsetVector);
+      } else {
+        final List<?> columns = row.getRow().get().getColumns();
+        Preconditions.checkNotNull(header, "Expected to read header row before any other data.");
+        rowQueue.add(new PullQueryRow(
+            columns,
+            expectedSchema,
+            row.getSourceHost(),
+            Optional.empty()
+        ));
```

Comment:
It appears that this is a stateful function from `StreamedRow` to `Optional<PullQueryRow>`.  

I'd be tempted to suggest not iterating over the rows and instead flatmapping over this function.  Should have the same outcome.  For me, not having the iterator in the way would help, but for others it make be more confusing...

Reply:
it needs to maintain state because rows can be added after its finished its first round of iteration but it needs to "remember" what it had done unless I'm missing what you're suggesting.

Reply:
The object `StreamedRowTranslator` can stay stateful (I mean, it must).  Instead of having methods which 1) add a list of rows, and 2) drain a list of results, one could have a single method which observes a row, and returns an `Optional<PullQueryRow>`.

Then instead of 

```      translator.add(data);
      while (translator.hasNext()) {
        queue.offer(new HandledRow(translator.next(), handler)); ...```
        
One would iterate over `data` and apply the function.       

The advantage is that one would be operating row by row instead of processing all the rows at once and building up another data structure.

Reply:
erm... that makes _a lot_ more sense than the way I have it. I think the way I have it is an artifact of some evil refactoring haha

Reply:
Made the change, it's now just:
```
for (final PullQueryRow row: translator.apply(data)) {
        queue.offer(new HandledRow(row, handler));
```